### PR TITLE
Add opt-in MPS profiling and reproducible benchmarks

### PR DESCRIPTION
**Pull request by enarjord**
## Summary

- make `PASSIVBOT_GPU_PROFILE=1` emit structured per-generation timing for proxy phases, NSGA ask/tell/orchestration, exact-validation queue/work, persistence, and checkpointing
- report dispatch chunks and kernel dispatches, actual batch sizes, topology-aware candidate-bars, candle/coin/side counts, optional metric features, and shared Metal-library cold/warm state
- add a bounded, deterministic, in-memory `passivbot tool gpu-proxy-benchmark` harness covering long-history single-coin EMA Anchor and Trailing Martingale, short-history multicoin overhead, and a coin-override case
- exercise the complete production proxy path in the harness, including full output transfer, metric reduction, and result materialization
- keep profiling disabled by default, keep ordinary CPU imports free of the optional MPS runtime, and keep exact Rust results authoritative

## Validation

- complete Python test suite: passed locally
- full Apple MPS suite: 717 passed
- opt-in real single-coin and multicoin proxy profiling tests: passed
- Rust: `cargo test --no-default-features` (274 passed)
- Rust: `cargo check --tests`
- AI docs checker: 0 errors (two pre-existing size warnings)
- focused CLI, GPU service/backend, benchmark, optimizer, shared-cache classification, topology/truncated-work accounting, nested-phase accounting, exact-worker timing, monotonic completion timing, GPU-specific dependency handling, and CPU no-MPS-import regression tests: passed

## Immutable benchmark comparison

Base `5009202751c55783d9440252588b158a06af7a39`; implementation `9b55f05a61b9409609f332de46bdb15f8d3e6866`. Each case ran in a fresh process through the complete production proxy evaluation path with seed 7, 256 candidates, five warm repeats, and matching fixture SHA-256 values. Single-coin cases use 60,000 bars; multicoin cases use 4,320 bars across eight coins. Warm p50 is separate from the first cold compilation run.

| case | base/head warm p50 cand/s | delta | base/head kernel p50 | base/head wall p50 |
|---|---:|---:|---:|---:|
| EMA single long | 372.23 / 372.29 | +0.01% | 0.64807 / 0.64950 s | 0.68774 / 0.68764 s |
| TM single long | 422.87 / 423.50 | +0.15% | 0.56251 / 0.56078 s | 0.60538 / 0.60449 s |
| EMA multicoin overhead | 496.00 / 490.57 | -1.10% | 0.47802 / 0.48333 s | 0.51613 / 0.52185 s |
| EMA multicoin overrides | 486.39 / 487.00 | +0.13% | 0.49121 / 0.48671 s | 0.52633 / 0.52567 s |

These measurements establish a reproducible baseline; the default path remains effectively flat within run-to-run noise, so this PR does not claim a speedup. Profiling synchronization is opt-in. No benchmark outputs, private configs, cache data, historical data, or optimization results are committed.

### CODE REVIEW

File: src/optimization/gpu/service.py [hunk 1]
```diff
@@ -109,6 +110,131 @@
...
+            (len(candidates) + dispatch_batch_size - 1) // dispatch_batch_size
+            if dispatch_batch_size > 0
+            else 0
+        ),
+        "actual_dispatch_batch_sizes": [],
+        "dispatch_count": 0,
+        "cold_dispatch_count": 0,
+        "warm_dispatch_count": 0,
+        "candidate_bars": len(candidates) * candle_count,
+        "kernel_candidate_bars": 0,
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/gpu/service.py]: Fixed in 3f7f3c4d: candidate_bars and kernel_candidate_bars now include candidates times candles times coins times sides. Single fused dual-side dispatches count both sides, while independent side dispatches naturally accumulate per dispatch. Focused topology-accounting coverage was added.

File: tests/optimization/test_gpu_proxy_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import numpy as np
+import pytest
+
+from optimization.gpu.model import (
+    EMA_ANCHOR_MULTICOIN_PARAM_KEYS,
+    EMA_ANCHOR_SINGLE_COIN_PARAM_KEYS,
+    TRAILING_MARTINGALE_SINGLE_COIN_PARAM_KEYS,
+)
+from tools.gpu_proxy_benchmark import (
```
[the remark this answers is not in the document]
Comment (enarjord) [reply tests/optimization/test_gpu_proxy_benchmark.py:11]: Fixed in a94a1129 and retained in 3f7f3c4d: optional MPS/torch imports are lazy and occur only when the benchmark executes. A CPU-only import regression test verifies collection does not require torch.

File: src/optimization/gpu/mps_kernel.py [hunk 3]
```diff
@@ -1266,7 +1267,8 @@ def run(self, params: np.ndarray, *, profile: bool = False) -> dict:
                 dtype=torch.int32,
                 device="mps",
             )
-        prepared = time.perf_counter()
+        prepared = time.perf_counter() if profile else 0.0
+        cold = self._profile_dispatches == 0
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/gpu/mps_kernel.py]: Fixed in 3f7f3c4d: cold/warm classification now reports an actual miss or hit in the shared LRU Metal-library cache via the cached-library helper, rather than runner age. Tests cover a cold first runner and a warm second runner sharing the cached library.

File: src/tools/gpu_proxy_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+            dtype=np.float32,
+        )
+        overrides[0, EMA_ANCHOR_COIN_OVERRIDE_WALLET_EXPOSURE_COLUMN] = 0.5
+        if coins > 1:
+            overrides[1, 0] = 0.04
+    runner = MpsEmaAnchorMulticoinRunner(
+        runs[0],
+        data,
+        side="long",
+        coin_overrides=overrides,
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/tools/gpu_proxy_benchmark.py:312]: Fixed in 3f7f3c4d: renamed the case to ema-multicoin-overhead. The requirement is a short-history multicoin overhead-sensitive workload; the production evaluation intentionally remains long-side only.

File: src/tools/gpu_proxy_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+    import torch
+
+    started = time.perf_counter()
+    output = runner.run(matrix, profile=True)
+    transfer_started = time.perf_counter()
+    for key in ("balance", "max_dd", "fill_count"):
+        value = output.get(key)
+        if value is not None:
+            value.detach().cpu()
+    device_to_host_seconds = time.perf_counter() - transfer_started
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/tools/gpu_proxy_benchmark.py]: Fixed in 3f7f3c4d: the harness now constructs lightweight production proxy objects and calls the production single-coin or multicoin evaluate method, covering complete output transfer, metric reduction, and result materialization. The immutable benchmark table was rerun through that path.

**Comment by enarjord** [comment]
@codex review

Please review exact head 3f7f3c4d4e08f1200246c643844d37c1c7d46c53. This head addresses all five prior findings: the short-history case is accurately named, the harness runs the complete production proxy path, optional torch imports remain lazy for CPU-only collection, candidate-bars include coin and side topology, and cold/warm compilation follows the shared Metal-library cache. The benchmark table has been rerun from immutable base and head commits.

File: src/optimization/gpu/service.py [hunk 6]
```diff
@@ -2876,6 +3165,20 @@ def evaluate(self, candidates: list[dict]) -> list[dict]:
...
+                extra_device_to_host = (
+                    profile["timings_seconds"]["device_to_host"]
+                    - stage_device_to_host_before
+                )
+                profile["timings_seconds"]["metric_reduction"] += (
+                    max(
+                        0.0,
+                        time.perf_counter()
+                        - stage_started
+                        - extra_device_to_host,
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/gpu/service.py]: Fixed in 7b29a68f: multicoin reduction accounting snapshots both device-transfer time and all runner-attributed phases before host reduction. Any HSL cutoff rerun packing, upload, compilation, kernel, decode, and transfer time is now subtracted from the enclosing elapsed interval, leaving only unattributed reduction work. A focused regression covers the residual calculation.

File: src/optimization/gpu/service.py [hunk 7]
```diff
@@ -109,6 +110,137 @@
...
+    profile["actual_dispatch_batch_sizes"].append(batch_size)
+    profile["dispatch_count"] += dispatch_count
+    profile["cold_dispatch_count"] += dispatch_count if cold else 0
+    profile["warm_dispatch_count"] += 0 if cold else dispatch_count
+    profile["kernel_candidate_bars"] += (
+        batch_size
+        * int(getattr(runner, "n", 0))
+        * int(getattr(runner, "n_coins", 1))
+        * int(side_count)
+        * dispatch_count
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/gpu/service.py:265]: Fixed in 7b29a68f: the cutoff replay forwards each candidate end step into runner-profile accounting. Kernel candidate-bars now sum the clipped effective steps and then apply coin and side topology, with focused tests for both truncated accounting and callback propagation.

File: src/optimization/gpu/service.py [hunk 8]
```diff
@@ -1718,29 +1864,72 @@ def _parameter_matrix(self, candidates: list[dict]) -> np.ndarray:
...
+                time.perf_counter() if self.profile_enabled else 0.0
+            )
             recovery_distribution = _mps_strategy_eq_recovery_distribution(
                 output, self.needed_metrics
             )
+            if profile is not None:
+                torch.mps.synchronize()
+                profile["timings_seconds"]["metric_reduction"] += (
+                    time.perf_counter() - stage_started
+                )
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/gpu/service.py:1973]: Fixed in 7b29a68f: the recovery-distribution helper now leaves its result on MPS. Profiling synchronizes after computation to close metric reduction, then the existing CPU copy occurs inside the device-to-host phase. The unit test now fails if the helper performs an early CPU copy.

**Comment by enarjord** [comment]
@codex review

Please review exact head 7b29a68fd815c58ed01051dce98ef64bedee511f. It addresses all three findings from the previous exact-head review: nested HSL replay phases are excluded from enclosing reduction time, truncated replay candidate-bars use effective per-candidate steps, and recovery-distribution copies are timed as device-to-host transfer. Focused tests and the full 717-test Apple MPS suite pass; the immutable benchmark table was rerun on this head.

File: src/tools/gpu_proxy_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+        parser.error("--coins must be at least two")
+    selected = CASES if args.case == "all" else (args.case,)
+    if any(name.endswith("single-long") for name in selected):
+        if candidates * single_bars > MAX_DISPATCH_CANDIDATE_BARS:
+            parser.error("single-coin candidate-bars exceed the safe benchmark limit")
+    if any(name.startswith("ema-multicoin") for name in selected):
+        if candidates * multicoin_bars * coins > MAX_DISPATCH_CANDIDATE_BARS:
+            parser.error("multicoin candidate-bars exceed the safe benchmark limit")
+
+    import torch
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/tools/gpu_proxy_benchmark.py]: Fixed in 525f9037: the benchmark now catches a missing torch package at the command boundary and exits through argparse with a concise installation hint for the full and gpu-mps extras. A CPU-safe regression verifies the message and exit code.

File: src/optimization/backends/gpu_backend.py [hunk 10]
```diff
@@ -4049,8 +4099,22 @@ def consume_ready(*, wait_for_one: bool = False) -> None:
...
+                float(payload.pop("__gpu_profile_worker_seconds__", 0.0))
+                if profile_enabled and isinstance(payload, dict)
+                else 0.0
+            )
+            if profile_enabled:
+                profile_totals["exact_work"] += worker_seconds
+                profile_totals["exact_queue_wait"] += max(
+                    0.0,
+                    result_ready_at - submitted_at - worker_seconds,
+                )
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/backends/gpu_backend.py]: Fixed in 525f9037: the parent samples submission time immediately before apply_async and passes it to the profiled worker. The worker records queue wait at its actual start boundary and records work duration separately, so ordered parent consumption and post-completion transfer delay cannot inflate queue wait. Focused tests pin both timestamps.

**Comment by enarjord** [comment]
@codex review

Please review exact head 525f9037316107a8cce2f22a30fe5db7ee031a59. It addresses both findings from the prior review: the public benchmark fails cleanly with an installation hint when the optional GPU package is absent, and exact-validation queue wait is recorded at the worker start boundary rather than parent result consumption. Focused tests pass, CI has been triggered, and the immutable benchmark table was rerun on this exact head.

File: src/passivbot_cli/main.py [hunk 11]
```diff
@@ -144,6 +144,11 @@ class CommandSpec:
         "tools.hsl_replay_benchmark",
         "benchmark the offline coin-HSL replay hot path",
     ),
+    "gpu-proxy-benchmark": CommandSpec(
+        "tools.gpu_proxy_benchmark",
+        "benchmark deterministic Apple MPS proxy workloads",
+        requires_full=True,
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/passivbot_cli/main.py]: Fixed in 9b55f05a: gpu-proxy-benchmark no longer uses the generic full-install gate. It reaches its own GPU-aware dependency check even when generic full markers are absent, so missing Torch produces the complete full plus gpu-mps installation guidance. The CLI regression now exercises that exact condition.

File: src/optimization/backends/gpu_backend.py [hunk 12]
```diff
@@ -4229,6 +4376,15 @@ def consume_ready(*, wait_for_one: bool = False) -> None:
         while pending and exact_done < budget:
             consume_ready(wait_for_one=True)
         maybe_save_checkpoint(force=True)
+        if profile_enabled:
+            _log_gpu_profile(
+                "complete",
+                generations=generation,
+                proxy_evaluations=proxy_evaluations,
+                exact_completed=exact_done,
+                wall_seconds=time.time() - start_time,
```
[the remark this answers is not in the document]
Comment (enarjord) [reply src/optimization/backends/gpu_backend.py]: Fixed in 9b55f05a: the optimizer captures a perf_counter start beside the ordinary logging start time, and the structured completion profile now derives wall_seconds only from that monotonic clock. A regression makes wall-clock access fail while verifying the elapsed helper.

**Comment by enarjord** [comment]
@codex review

Please review exact head 9b55f05a61b9409609f332de46bdb15f8d3e6866. It addresses both findings from the prior review: gpu-proxy-benchmark now reaches its GPU-specific dependency gate instead of the generic full-install preflight, and the structured completion wall duration uses perf_counter rather than wall-clock time. Focused regressions pass, CI has been triggered, and the immutable benchmark table was rerun on this exact head.
